sqlglot/optimizer/qualify_columns.py

```diff
@@ -380,18 +410,89 @@ def _qualify_columns(scope: Scope, resolver: Resolver) -> None:
                     column.set("table", column_table)
 
 
+def _expand_struct_stars(
+    expression: exp.Dot,
+    next_alias_name: t.Callable[[], str],
+) -> t.List[exp.Alias]:
+    struct_fields: t.List[str] = []
+
+    # [BigQuery] Expand/Flatten foo.bar.* where bar is a struct column
+    starting_col = expression.find(exp.Column)
+    assert starting_col
+
+    dot_parts = expression.parts
+    table_name = starting_col.table
+
+    # If there's more nested levels specified, start the DFS traversal from the nested struct first
+    for part in dot_parts[2:-1]:
+        datatype = starting_col.find(exp.DataType) or (
+            starting_col.type and starting_col.type.find(exp.DataType)
+        )
+        assert datatype
+
+        for expr in datatype.expressions:
+            if not isinstance(expr.this, exp.Identifier):
+                return []
+
+            if expr.this.this == part.this:
+                starting_col = expr
+                assert starting_col
+
+    if not starting_col or starting_col.this.this != dot_parts[-2].this:
+        raise OptimizeError(f"Unknown struct: {part.this}")
+
+    starting_name = ".".join([part.this for part in dot_parts[1:-2]])
+    stack = [(starting_col, starting_name)]
+
+    # Traverse & flatten the starting_col struct
+    while stack:
+        col, path_name = stack.pop()
+        assert col
+
+        if not isinstance(col.this, exp.Identifier):
+            return []
```

Comment:
Ditto about checking for `ColumnDef`.

Reply:
Ditto about the assertion of `Column | ColumnDef`